psl/schema-ast/src/ast/find_at_position/attribute.rs

```diff
@@ -9,50 +9,60 @@ pub enum AttributePosition<'ast> {
-            // between arguments and that's hard to capture in the pest grammar.
-            let mut spans: Vec<(Option<&str>, ast::Span)> = attr
-                .arguments
-                .iter()
-                .map(|arg| (arg.name.as_ref().map(|n| n.name.as_str()), arg.span()))
-                .chain(
-                    attr.arguments
-                        .empty_arguments
-                        .iter()
-                        .map(|arg| (Some(arg.name.name.as_str()), arg.name.span())),
-                )
-                .collect();
-
-            spans.sort_by_key(|(_, span)| span.start);
-
-            let mut arg_name = None;
-            for (name, _) in spans.iter().take_while(|(_, span)| span.start < position) {
-                arg_name = Some(*name);
-            }
+        let mut spans: Vec<(Option<&str>, ast::Span)> = attr
+            .arguments
+            .iter()
+            .map(|arg| (arg.name.as_ref().map(|n| n.name.as_str()), arg.span()))
+            .chain(
+                attr.arguments
+                    .empty_arguments
+                    .iter()
+                    .map(|arg| (Some(arg.name.name.as_str()), arg.name.span())),
+            )
+            .collect();
 
-            // If the cursor is after a trailing comma, we're not in an argument.
-            if let Some(span) = attr.arguments.trailing_comma {
-                if position > span.start {
-                    arg_name = None;
-                }
+        spans.sort_by_key(|(_, span)| span.start);
+
+        let mut arg_name = None;
+        for (name, _) in spans.iter().take_while(|(_, span)| span.start < position) {
+            arg_name = Some(*name);
+        }
+
+        // If the cursor is after a trailing comma, we're not in an argument.
+        if let Some(span) = attr.arguments.trailing_comma {
+            if position > span.start {
+                arg_name = None;
             }
+        }
 
-            if let Some(arg_name) = arg_name.flatten() {
-                return Self::Argument(arg_name);
+        if let Some(arg) = attr.arguments.iter().find(|arg| arg.span().contains(position)) {
+            if let ExpressionPosition::FunctionArgument(fun, name) = ExpressionPosition::new(&arg.value, position) {
+                return Self::FunctionArgument(fun, name, arg.value.to_string());
             }
 
-            if let Some(arg) = attr.arguments.iter().find(|arg| arg.span().contains(position)) {
-                if let ExpressionPosition::FunctionArgument(fun, name) = ExpressionPosition::new(&arg.value, position) {
-                    return Self::FunctionArgument(fun, name);
+            if arg.value.is_array() {
+                let arr = arg.value.as_array().unwrap();
+                let expr = arr.0.iter().find(|expr| expr.span().contains(position));
+                if let Some(expr) = expr {
+                    return Self::ArgumentValue(arg.name(), expr.to_string());
```

Comment:
Problem: when using nested composite type fields in attributes, i.e.
```prisma
model User {
    id      String  @id @map("_id") @db.String
    address Address

    @@unique([address.poBox.name])
}

type Address {
    street String  @db.String
    city   String
    poBox    POBox
}

type SubType {
    name String
}
```

We retrieve the whole string that is defined, i.e.
```
"address.poBox.name"
```

and dont have more granular understanding of if the cursor is inside `address`, `poBox` or `name`.

This PR already supports finding fields from non-nested fields in block attributes, but I'm not sure what to do with this.

This seems like it would require modification to this pest grammar:

```
expression = { function_call | array_expression | numeric_literal | string_literal | path }

string_literal = ${ "\"" ~ string_content ~ "\"" }

string_content = @{ (string_escape | !("\"" | ASCII_CONTROL_CHARACTER) ~ ANY)* }
```

But I'm not fully sure of the ramifications of changing the pest grammar like this. If this is something we want to explore, I think it should be in a follow-up PR as a stretch goal

Reply:
The string literal rule is irrelevant here, it matches a quoted string literal in source code of the schema and there are no quotes here. The relevant rule is `path`, and it is already defined as

```pest
identifier = @{ ASCII_ALPHANUMERIC ~ ( "_" | "-" | ASCII_ALPHANUMERIC)* }
path = @{ identifier ~ ("." ~ path?)* }
```

so no modifications are necessary in the grammar, it is already defined the way you need.

You need to modify the `Rule::path` case in `parse_expression` in `psl/schema-ast/src/parser/parse_expression.rs` to recurse into `path` and create a separate span for each `identifier` token instead of converting the whole subtree to string as it does now.